buidl/script.py

```diff
@@ -369,34 +371,92 @@ def convert(cls, raw_redeem_script):
+
+        For safety, you may pass in an expected_addr and the method will throw an error if the derived address doesn't match the expected one.
         """
         # safety checks
         if type(quorum_m) is not int:
             raise ValueError(f"quorum_m must be of type int: {quorum_m}")
-        if quorum_m < 1 or quorum_m > len(pubkey_hex_list):
-            raise ValueError(f"Invalid m-of-n: {quorum_m}-of-{len(pubkey_hex_list)}")
+        if quorum_m < 1 or quorum_m > len(pubkey_hexes):
+            raise ValueError(f"Invalid m-of-n: {quorum_m}-of-{len(pubkey_hexes)}")
 
         commands = [number_to_op_code(quorum_m)]
 
         if sort_keys:
-            pubkey_hexes = sorted(pubkey_hex_list)
+            pubkey_hexes = sorted(pubkey_hexes)
         else:
-            pubkey_hexes = pubkey_hex_list
+            pubkey_hexes = pubkey_hexes
 
         for pubkey_hex in pubkey_hexes:
             # we want these in binary (not hex)
             commands.append(bytes.fromhex(pubkey_hex))
 
-        quorum_n = len(pubkey_hex_list)
+        quorum_n = len(pubkey_hexes)
         commands.append(number_to_op_code(quorum_n))
         commands.append(174)  # OP_CHECKMULTISIG
 
-        return cls(commands)
+        to_return = cls(commands)
+
+        if expected_addr:
+            calculated_addr = to_return.address(network=expected_addr_network)
+            if expected_addr != calculated_addr:
+                raise ValueError(
+                    f"Expected address {expected_addr} but calculated {calculated_addr}"
+                )
+
+        return to_return
+
+    @classmethod
+    def create_p2sh_multisig_unsorted(
+        cls, quorum_m, pubkey_hexes, target_address, network="main"
+    ):
+        """
+        Helper method to create a p2sh that orders pubkeys to get a target_address.
+
+        DO NOT USE THIS METHOD IF YOU DON'T UNDERSTAND WHAT IT'S DOING.
+        THIS SHOULD ONLY BE USED FOR SPENDING OUT OF A LEGACY ADDRESS.
+        DO NOT SEND NEW FUNDS TO AN ADDRESS GENERATED USING THIS METHOD.
+        Instead, use the regular create_p2sh_multisig method (with sort_keys=True) to generate a modern address.
+        """
+        for permutation in permutations(pubkey_hexes):
+            result = cls.create_p2sh_multisig(
+                quorum_m=quorum_m, pubkey_hexes=permutation, sort_keys=False
+            )
+            if result.address(network) == target_address:
+                return result
+
+        # We didn't get a match
+        raise ValueError(
+            f"Could not find a combination of pubkeys the generated address {target_address}"
+        )
+
+    def get_quorum(self):
+        """
+        Return the m-of-n of this multisig, as in 2-of-3 or 3-of-5
+        """
+        if self.commands[-1] != 174:
+            raise ValueError(f"Not OP_CHECKMULTISIG: {self}")
+        quorum_m = op_code_to_number(self.commands[0])
+        quorum_n = len(self.commands) - 3
```

Comment:
A comment about where the `3` comes from would be nice (I think it's `OP_CHECKMULTISIG`, `quorum_m` and the off-by-one issue?)

Reply:
Correct! Added a comment to make it unambiguous in https://github.com/buidl-bitcoin/buidl-python/pull/81/commits/5c3bd5ecff35bf0b38e2523c60d6b218e11c61ea